// mage-finger/lib/collector.js

var url = require('url')
    , http = require('http');

http.Agent = 'Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_5_8; en-US) AppleWebKit/532.5 (KHTML, like Gecko) Chrome/4.0.249.0 Safari/532.5';
//console.log('Using user agent "' + http.Agent + '"');

exports.version = '0.1.0';

exports.collector = function(target, options) {
    this.target = target;
    this.options = options;
    var results = [];

    // List all tests here
    this.fingers = [ 'release-notes', /*'dummy1', 'dummy2',*/ 'req-by-module' ]

    // Run tests
    this.finger = function() {
        if (! this.target.match(/^https?:\/\/[a-z0-9][a-z0-9.-]*\.[a-z]{2,20}/i)) {
            throw new Error('Invalid target base URL: ' + this.target);
        }
        var target = url.parse(this.target);

        for (var i in this.fingers) {
            var code = this.fingers[i];
            // bind method to this instance using a closure
            require('./fingers/' + code).run(target, (function(closure){
                return function(){closure._testComplete.apply(closure, arguments)}
            })(this));
        }
    }

    // Internal callback when a test is completed
    this._testComplete = function(finger, versions) {
        results.push({'code': finger, 'versions': versions});
        if (typeof this.options.testcomplete == 'function') {
            this.options.testcomplete(finger, versions);
        }

        if (results.length == this.fingers.length) {
            this._allComplete(results);
        }
    }

    // Internal callback when all tests are completed
    this._allComplete = function(results) {

        var scoreLimit = this.options.scoreLimit ? this.options.scoreLimit : 1;
        var resultLimit = this.options.resultLimit ? this.options.resultLimit : 0;

        var results = _processResults(results, scoreLimit, resultLimit);
        if (typeof this.options.allcomplete == 'function') {
            this.options.allcomplete(results);
        }
    }

    // Internal method to summarize and sort the results
    var _processResults = function(results, scoreLimit, resultLimit) {
        var tmp = {};
        var bymodule = {};
        var code, version;
        if (results.length) {
            for (var i = 0; i < results.length; i++) {
                code = results[i].code;
                if (results[i].versions && results[i].versions.length) {
                    for (var j = 0; j < results[i].versions.length; j++) {
                        version = results[i].versions[j];
                        if (typeof tmp[version] == 'undefined') {
                            tmp[version] = 0;
                        }
                        if (typeof bymodule[version] == 'undefined') {
                            bymodule[version] = [];
                        }
                        tmp[version]++;
                        bymodule[version].push(code)
                    }
                }
            }
        }

        var out = [];
        var total = 0;
        // Filter out results that don't match the scoreLimit
        for (version in tmp) {
            if (scoreLimit > tmp[version]) continue;
            total += tmp[version];
            out.push({'version': version, 'score': tmp[version], 'tests': bymodule[version]});
        }

        // Sort results
        out = out.sort(_sortResults);

        // Cut off results beyond the resultLimit
        if (resultLimit && out.length > resultLimit) {
            out = out.slice(0, resultLimit);
            total = 0;
            for (i in out) {
                total += out[i].score;
            }
        }

        // Add percentages on remaining entries
        for (i in out) {
            out[i].percent = 100 / total * out[i].score;
        }

        return out;
    }

    var _sortResults = function(a, b)
    {
        if (a.score == b.score) {
            if (a.version == b.version) return 0;
            if (a.version > b.version) return -1
            return 1;
        }
        if (a.score > b.score) return -1;
        return 1;
    }
}
