// mage-finger/lib/collector.js

var url = require('url');

exports.version = '0.1.0';

exports.collector = function(target, options) {
    this.target = target;
    this.options = options;
    var results = [];

    // List all tests here
    this.fingers = {
        'releasenotes': require('./release-notes')
    }

    this.finger = function() {
        var target = url.parse(this.target);

        for (var finger in this.fingers) {
            this.fingers[finger].run(target, (function(closure){
                return function(){closure._testComplete.apply(closure, arguments)}
            })(this));
        }
    }

    this._testComplete = function(finger, versions) {
        results.push({'code': finger, 'versions': versions});
        if (typeof this.options.testcomplete == 'function') {
            this.options.testcomplete(finger, versions);
        }

        if (results.length == Object.keys(this.fingers).length) {
            this._allComplete();
        }
    }

    this._allComplete = function() {
        if (typeof this.options.allcomplete == 'function') {
            this.options.allcomplete(results);
        }
    }
}
